### ISSUE TITLE
support vte 0.38 in gtk3 build

### ISSUE THREAD

**Issue by dbevans** [geany/geany#336]
As of vte 0.37, API version has been incremented to 2.91, and, as of vte 0.38, a number of API changes have been introduced which break vte support in the gtk3 build.

In particular, deprecated function vte_terminal_fork_command() has been removed and replaced by vte_terminal_spawn_sync() (previously vte_terminal_fork_command_full()).

vte 0.38 has been released concurrently with GNOME 3.14 which depends upon it.

**Comment by egmontkob** [comment]
Just FYI:

Vte-0.40 brings back the word-chars support (with a slightly different API) due to popular demand.

It also adds an API for getting the runtime version.  It doesn't solve the GdkColor vs GdkRGBA issue right now, but might come handy in the future. You can safely detect >=0.40 from now on, and you only need some nasty temporary hack to distinguish 0.38 from <=0.36 (e.g. if getting the version is not yet available and set_emulation is no longer there then it's 0.38... or something like this).

**Comment by b4n** [comment]
> Vte-0.40 brings back the word-chars support (with a slightly different API) due to popular demand.

Okay, that'd be usable.

> It also adds an API for getting the runtime version.

How so?  I can't seem to find it in the docs?

**Comment by egmontkob** [comment]
On Sat, Jul 4, 2015 at 3:26 PM, Colomban Wendling <notifications@github.com>
wrote:

> Vte-0.40 brings back the word-chars support (with a slightly different
> API) due to popular demand.
>
> Okay, that'd be usable.
>
> It also adds an API for getting the runtime version.
>
> How so? I can't seem to find it in the docs?
>
> See vte_get_{major,minor,micro}_version() in the source or the header
files.  They're indeed missing from the docs, thx for pointing it out, I'll
fix it.

e.

**Comment by Akronix** [comment]
hi!
How is this going? any new about this??

**Comment by b4n** [comment]
No.  Maybe the latest version could be usable, I'd have to check.  For the moment just use 2.90.

**Comment by egmontkob** [comment]
FYI: Debian and Ubuntu have already dropped libvte 2.90 (tarball 0.36). They only ship the ancient gtk2 version (tarball 0.28), as well as 2.91 for gtk3 (tarball 0.38 or newer).

Neither the gtk2 version, nor 2.90 is maintained anymore by upstream. Bugfixes / security fixes / improvements only go to the newest tarballs (library version 2.91).

The first version with the new API was released almost 2 years ago. I have myself ported a vte-based terminal emulator (whose codebase I was not familiar with), it took maybe like 2-3 hours, can't remember, but definitely not much more.

So I guess it's high time port geany to the newest vte. (I don't think parallel support for 2.90 and 2.91 makes any sense, just go for 2.91 and forget 2.90 for good).

A note about `vte_terminal_set_background_image_file()`: Indeed direct support for background image has been removed; however, see the Terminix project how a picture can be drawn behind a transparent/translucent VTE using some GTK+ magic.

**Comment by b4n** [comment]
> The first stable version with the new API was released almost 2 years ago. I have myself ported a vte-based terminal emulator (whose codebase I was not familiar with), it took maybe like 2-3 hours, can't remember, but definitely not much more.

The problem is that for some reason we `dlopen()` VTE, and 0.38 lacks runtime version checking, so it's harder than it should.  But yeah maybe we could play and assume it's 0.38 if an older symbol is missing and a newer present (although it's made even more annoying to do than it should because the [0.38 docs](https://developer.gnome.org/vte/0.38/) lacks any "new in 0.38" docs, so it has to be guessed).
It'd probably be doable, but VTe really doesn't help here.

> (I don't think parallel support for 2.90 and 2.91 makes any sense, just go for 2.91 and forget 2.90 for good).

No.  Unless there's a compelling reason to drop 2.90, we should not drop it.

> A note about `vte_terminal_set_background_image_file()`: Indeed direct support for background image has been removed; however, see the Terminix project how a picture can be drawn behind a transparent/translucent VTE using some GTK+ magic.

Probably makes sense.  But well, it's a retarded feature anyway, so why bother.

**Comment by egmontkob** [comment]
> because the 0.38 docs lacks any "new in 0.38" docs, so it has to be guessed

The doc is designed for people writing new code against VTE, not for those porting their code from older versions. As such, it solely documents the 2.91 API, and therefore only mentions additions that weren't yet available in the very first 2.91 version (that is, things you need to be aware of if developing against 2.91). In other words, on a backwards incompatible API bump the doc is started over from scratch. With this regard, it's absolutely irrelevant when something appeared and how (if differently) looked like in the previous API. (I do not intend to discuss whether it's good or bad this way, I just stated that this is the reason. I think it's the same with, let's say, GTK+ version 3, you won't see any GTK+-2 stuff or 2->3 porting advice mentioned in GTK+-3's docs.)

> No. Unless there's a compelling reason to drop 2.90, we should not drop it.

Con of dropping: Save one step for users who have an old enough system not to have libvte 2.91, however, wish to have the latest Geany with as little hassle as possible. (I never quite understood this goal, I believe those wishing to run the latest apps are expected to keep their base system reasonably up to date as well. Sure it's nice to have some window, but maybe 2 years should be enough.)

Pros of dropping: Cleaner, simpler code, easier maintenance, fewer setups to test. Plus, the aforementioned users, although suffer from the inconvenience of having to install another lib, will get the new features, bufixes, security/privacy fixes (e.g. scrollback buffer no longer written plaintext to disk) of new vte.

As a VTE developer, I see no reason in supporting both versions in parallel, I pretty much find it a waste of the precious (and apparently scarce) developer resources. However, your project, your call.

**Comment by b4n** [comment]
> The doc is designed for people writing new code against VTE […]

Ok, mostly makes sense.

On this topic though, the docs are somewhat broken.  See for example [the 0.38 docs for `vte_terminal_set_color_foreground()`](https://developer.gnome.org/vte/0.38/VteTerminal.html#vte-terminal-set-color-foreground): it doesn't have the proper prototype.  [0.44 ones](https://developer.gnome.org/vte/0.44/VteTerminal.html#vte-terminal-set-color-foreground) have it, but lost the description.

> Plus, the aforementioned users, although suffer from the inconvenience of having to install another lib, will get the new features, bufixes, security/privacy fixes (e.g. scrollback buffer no longer written plaintext to disk) of new vte.

That only depends on us supporting the newer API and the user having that newer lib.  The only advantage of dropping 2.90 support would be for users to get no VTE support and complain, then tell them to install 2.91, instead of them not seeing they miss 2.91 but have 2.90.  Nah, we're not babysitters, they can do it.
And yeah, I myself don't care support for older systems, I use a recent one.  But we always have kept support for older systems whenever it wasn't too crazy, and I won't change that just for VTE 0.38.

> As a VTE developer, I see no reason in supporting both versions in parallel […]

The VTE developers surely won't want to support 2.90 anymore, that's fairly understandable.  Though I wish they continued on a GTK2-compatible version, but well :)

**Comment by egmontkob** [comment]
> 0.44 ones have it, but lost the description

Thanks, filed https://bugzilla.gnome.org/show_bug.cgi?id=770114. For the time being, you can consult the docs next to the methods in the source code.

Feel free to file any further bugs you encounter to our bugtracker!

> Though I wish they continued on a GTK2-compatible version

We're also running on very limited developer resources. Every minute spent on maintaining the GTK2 version would have been taken away from making the GTK3 version better, effectively resulting in the world having a significantly worse GTK3-based VTE today. I'm absolutely convinced ditching GTK2 support was the best we could do.

**Comment by elextr** [comment]
> We're also running on very limited developer resources. Every minute spent on maintaining the GTK2 version would have been taken away from making the GTK3 version better, effectively resulting in the world having a significantly worse GTK3-based VTE today. I'm absolutely convinced ditching GTK2 support was the best we could do.

Pity about the projects using it, now they all have to spend the effort to update.

@b4n, just take VTE out of Geany until it gets its shit together, simple solution.

**Comment by b4n** [comment]
https://github.com/b4n/geany/commits/vte-2.91 seems to work, but for background images.  I can't seem to find a non-crazy way of implementing it (I can't get CSS to do it), and I don't care enough about it to implement a crazy way.  But we still need a proper thing here, either implementing it somehow, or dropping the feature altogether (I'd vote for the second, but well).

**Comment by codebrainz** [comment]
> But we still need a proper thing here, either implementing it somehow, or dropping the feature altogether (I'd vote for the second, but well).

If we can just guard out the feature for unsupported versions, that would be cool. Anyone who wants support with VTE that doesn't natively support the feature anymore can give patches to VTE project or else try and get a "workaround" PR merged into Geany.

**Comment by b4n** [comment]
it's probably possible to hide the pref UI when the VTE API is missing, yeah.

**Comment by codebrainz** [comment]
Yeah, that's what I meant. IIRC the patch that introduced the feature was fairly small.

**Comment by egmontkob** [comment]
@elextr OMG you need to update your code *once*!!! Which, as I've said, would take maybe a few hours, and the same goes for about a dozen projects out there in the wild using VTE. Compare this to the amount of time and effort it would have taken us to develop and test each and every change (hundreds of them) during the years for two different GTK+ versions in parallel. No one can seriously think that that would have been the better approach.

I'm here absolutely voluntarily to help you out to get Geany ported to the newer version, and also give you rationale behind our decisions, and up to now I was happy to do it.

I've contributed *a lot* voluntarily to VTE (and *many* VTE-based apps) and I truly enjoyed it! It doesn't take a genius to realize what I've already mentioned: Had we had to maintain VTE for two different versions in parallel, we would have had less time to develop the GTK+-3 version and hence currently it wouldn't be anywhere near as good as it is now. But there's one more factor to this: I did it happily and passionately as a hobby project. Had I had to maintain an ancient version too in parallel to the current one, I wouldn't have been that happy and that passionate, and hence would have spent an overall significantly smaller time on the project.

I've faced disagreements, conflicts, disappointments throughout these times, but never during my volunteered contributions for these years have I ever felt offended – but I do now, by you calling our work "shit".

You know what... pretty much every other VTE-based project has already managed to update their code by now, Geany is one of the very few left behind... don't you think that maybe YOU should get YOUR shit together?!

And you owe us (VTE developers) an apology!

**Comment by elextr** [comment]
@egmontkob basically everything you say applies to Geany as well, we are all volunteers who do Geany in our own time.  

And as we have to still support GTK2 VTE as well, we have to have #idefs or runtime tests to handle changes of API through time, the very thing you are refusing to do.  And that is keeping us from doing useful things with Geany.

As for removing VTE, well, I have been saying that for ages, since way before GTK3, so thats nothing new.

**Comment by egmontkob** [comment]
@elextr 

> ... as we have to still support GTK2 ...

No, you don't _have to_. Your project has decided to support it rather than sticking to a particular GTK and spending the freed up resources on something else. It was your project's decision, and...

> And that is keeping us from doing useful things with Geany.

... apparently you yourself are in doubt whether it was a good one.

So just because VTE gave the other possible answer to the same question, why does it make our project shit???

You still owe me that apology for your unacceptable style (no, not the different opinion, you're free to have that; it's the style), and until I receive that, I'm off from this discussion.

**Comment by elextr** [comment]
> why does it make our project shit???

It doesn't, and I didn't say it did, [get your shit together](http://www.urbandictionary.com/define.php?term=get%20your%20shit%20together) is simply a phrase meaning to get organised.  It expressed my frustration at unstable APIs.

**Comment by b4n** [comment]
@elextr to be fair, they broke API and that's all.  It's perfectly understandable, and when it doesn't happen every 5th of the month, it's generally warranted by good reasons -- heh, we even do that ourselves from time to time :)  E.g., switching a GTK3 API to use GdkRGBA *is* sensible per se, they just happened to keep better-than-sane compatibility in the 2.90 era.  That's unfortunate for us to some extent (as we could have assumed new API if GTK3), but we can hardly blame them for that.

The only point where I actually don't like 0.38 is that it doesn't have proper runtime version checks -- which were introduced later on --, yet our use somewhat requires it.  But well, that's only relevant to us because for some reason we `dlopen()` VTE instead of properly linking to it and making a build-time decision; and we're likely the odd kid here.

**Comment by elextr** [comment]
> ... instead of properly linking to it and making a build-time decision; and we're likely the odd kid here.

Yeah, @codebrainz has persuaded me that we could just drop that runtime stuff now, but we still need various `#ifs` to handle all the various versions IIUC.

**Comment by b4n** [comment]
yes, but kinda like we have for GTK3, not really any more or less.

**Comment by elextr** [comment]
GTK3, not a good example to use for me ;-)

**Comment by codebrainz** [comment]
All the `#if` stuff could be isolated in a single module (vte.c), and not even spread out as much as the current runtime `dlopen()` support (we currently check at compile-time _and_ runtime for much stuff).

BTW, I'm currently tinkering with geany-vte.m4 to try and handle all the VTE versions including with and without gtk3, and whether `--enable-vte` was passed yes or auto, it's a combinatorial explosion, but it would make the code nicer/easier.

**Comment by Akronix** [comment]
Hi!

I just want to say two things:
1) I pushed the issue up because I was using Geany with `enable-gtk3` activated and I was missing the VTE (*I do use it a lot*). I'm on Ubuntu 16.04 and as @techee said in #1181 :
> Ubuntu 16.04 has only vte 2.91 for GTK3.

2) I didn't like how some of the Geany devs deal with the discussion from the beginning. It's not something that invite to participate or contribute to read those lines; I think that a nicer environment is much more productive and helper for everyone to come in.
Just I'd ask to be more easygoing for the next time.

**Comment by elextr** [comment]
@Akronix yeah, maybe it was an overreaction on both sides, but as #1181 shows its not a trivial change and effort we could productively use elsewhere.

**Comment by Akronix** [comment]
I see, but I don't find it as a superfluous effort at all.

Either way, thank you @b4n for your work.
